talk/morelanguage/exceptions.tex

```diff
@@ -4,185 +4,166 @@
+        if (fh = open_file(file)) {
+          throw
+            file_bad(file.error.reason());
+        }
+        while (!fh) {
+          line = read_line(fh);
+          add_line_to_db(line); // Can throw
+                                // db_bad
         }
+        return 0;
       }
     \end{cppcode*}
   \end{multicols}
 \end{frame}
 
-\begin{frame}[fragile]
-  \frametitlecpp[98]{Controlling exceptions \hfill \deprecated/\removed}
-  \begin{block}{Declaring expected exceptions}
-    \begin{itemize}
-    \item each function can declare a set of expected exceptions
-    \item using the {\it throw} statement in its declaration
-    \item other exceptions won't exit the scope of the function
-      \begin{itemize}
-      \item instead, the {\it unexpected} handler is called
-      \item by default, it terminates the program
-      \end{itemize}
-    \end{itemize}
-  \end{block}
-  \pause
-  \begin{cppcode}
-    int func(int a) throw(int) {
-      if (0 == a) {
-        throw 2;  // ok, goes out
-      } else {
-        throw "hello"; // std::unexpected called
-      }
-    }
-  \end{cppcode}
-\end{frame}
-
-\begin{frame}[fragile]
-  \frametitlecpp[98]{Controlling exceptions \hfill \deprecated/\removed}
-  \begin{block}{Good to know}
-    \begin{itemize}
-    \item The check is done at runtime, not at compile time
-      \begin{itemize}
-      \item unlike Java
-      \end{itemize}
-    \item When the {\it throw} clause is absent, any exception can go out
-    \item To block all exceptions, use {\it throw()}
-    \end{itemize}
-  \end{block}
-  \pause
-  \begin{cppcode}
-    int func(int a) {
-      // any exception can go out
-    }
-    int otherfunc(int a) throw() {
-      // no exception can go out
-    }
-  \end{cppcode}
-\end{frame}
 
 \begin{frame}[fragile]
-  \frametitlecpp[11]{Removal of \cpp98 exceptions}
+  \frametitlecpp[98]{Error Handling and Exceptions}
   \begin{block}{}
-    After a lot of thinking and experiencing, the conclusions of the community on exception handling are :
     \begin{itemize}
-    \item Never write an exception specification
-    \item Except possibly an empty one
-    \end{itemize}
-  \end{block}
-  \pause
-  \begin{alertblock}{Some of the reasons}
-    \begin{itemize}
-    \item throw specification is runtime only
+      \item exceptions have little cost if no exception is thrown
       \begin{itemize}
-      \item does not allow compiler optimizations
-      \item on the contrary forces extra checks
-      \item generally terminates your program if violated
+        \item a throw usually costs about as much as a function call
+        \item so minimal if you are using exceptions for \textit{exceptional} events
+        \item thus their use is generally recommended
       \end{itemize}
-    \item throw specification clashes with templates
+      \item however, in a tight, performance critical loop, where the error is handled close to the place it happens a C-style error code might be better
```

Comment:
I would not recommend solely a C-style error code. We have `boost::outcome`, `std::expected`, `std::system_error` or just a plain sentinal value (e.g. returning -1 if the normal return values are expected to be positive). Also C-style error code could imply `errno` here, which is a `thread_local` variable and that is also expensive.

Reply:
@graeme-a-stewart Maybe you missed this comment.

Reply:
Fair point - now I say more explicitly "error code or object". I am a bit reluctant to get into the details of all the options in this slide (after all, it's the "exception" to exceptions, 😄).
- `boost::outcome` is out of scope for a core C++ course (Boost isn't really mentioned at all)
- `std::expected` didn't make it to C++20 AFAICS
- `std::system_error` is a lot heavier than an `int` (this is a tight loop, by definition)